### PR TITLE
Entity list

### PR DESCRIPTION
**Pull request by wpride**
Contains https://github.com/dimagi/formplayer/pull/28/files

This PR adds new objects and commands for getting the case list and details data in the format that we need

Also adds an entire application for testing, hence the large amount of files changed

### CODE REVIEW

File: gradlew [hunk 1]
```diff
@@ -42,6 +42,11 @@ case "`uname`" in
     ;;
 esac
 
+# For Cygwin, ensure paths are in UNIX format before anything is touched.
```
Comment (benrudolph) [inline gradlew]: haha how did you run into this?

File: src/main/java/application/MenuController.java [hunk 2]
```diff
@@ -65,10 +70,38 @@ public SessionBean performInstall(@RequestBody InstallRequestBean installRequest
...
+     * Make a a series of menu selections (as above, but can have multiple)
+     *
+     * @param menuSelectRepeater Give the selections to be made on the current MenuSession
+     *                             (could be a module, form, or case selection)
+     * @return A MenuResponseBean or a NewFormSessionResponse
+     * @throws Exception
+     */
+    @ApiOperation(value = "Make the given menu selection and return the next set of options, or a form to play.")
+    @RequestMapping(value = Constants.URL_MENU_SELECT_REPEATER, method = RequestMethod.POST)
+    public SessionBean selectMenuRepeater(@RequestBody MenuSelectRepeater menuSelectRepeater) throws Exception {
```
Comment (benrudolph) [inline src/main/java/application/MenuController.java]: a lot of this logging seems like it should be `debug` not `info`, but don't have too strong of feelings

File: src/main/java/application/MenuController.java [hunk 3]
```diff
@@ -87,34 +120,42 @@ private SessionBean getNextMenu(MenuSession menuSession, boolean redrawing) thro
...
+                menuResponseBean = generateMenuScreen((MenuScreen) nextScreen);
             }
             // We're looking at a case list or detail screen (probably)
             else if (nextScreen instanceof EntityScreen) {
-                EntityScreen entityScreen = (EntityScreen) nextScreen;
-                menuResponseBean.setMenuType(Constants.MENU_ENTITY);
-                menuResponseBean.setOptions(getMenuRows(entityScreen.getCurrentScreen()));
+                menuResponseBean = generateEntityScreen((EntityScreen) nextScreen);
+            } else{
+                throw new Exception("What screen are we on? " + nextScreen);
```
Comment (benrudolph) [inline src/main/java/application/MenuController.java]: nit: might be an easier error to read if it's `Unable to determine next screen: `

File: src/main/java/application/MenuController.java [hunk 4]
```diff
@@ -87,34 +120,42 @@ private SessionBean getNextMenu(MenuSession menuSession, boolean redrawing) thro
...
+        return new EntityListResponse(nextScreen);
+    }
     private MenuSession getMenuSession(String sessionId) throws Exception {
         return new MenuSession(menuRepo.find(sessionId), restoreService, installService);
     }
+
+    @ExceptionHandler(Exception.class)
+    public String handleError(HttpServletRequest req, Exception exception) {
+        log.error("Request: " + req.getRequestURL() + " raised " + exception);
+        exception.printStackTrace();
```
Comment (benrudolph) [inline src/main/java/application/MenuController.java]: this still feels weird to me, but i don't really know how to better handle errors in java

File: src/main/java/auth/BasicAuth.java [hunk 5]
```diff
@@ -29,4 +29,9 @@ public HttpHeaders getAuthHeaders() {
             }
         };
     }
+
+    @Override
+    public String toString(){
+        return "BasicAuth [username=" + username + ", password=" + password + "]";
```
Comment (benrudolph) [inline src/main/java/auth/BasicAuth.java]: would this object every get `toString`'d and sent somewhere where it shouldn't? like email or something? if so we should remove `password` from this

Comment (benrudolph) [inline src/main/java/auth/BasicAuth.java]: also i tend to like this `toString` format best, would be awesome if it was consistent throughout

File: src/main/java/beans/menus/EntityDetailResponse.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+ */
+public class EntityDetailResponse extends MenuSessionBean{
+    private Object[] details;
+    private Style[] styles;
+    private String[] headers;
+
+    public EntityDetailResponse(){}
+
+    public EntityDetailResponse(EntityDetailSubscreen entityScreen){
+        //TODO WSP Duh
```
Comment (benrudolph) [inline src/main/java/beans/menus/EntityDetailResponse.java]: ?

File: src/main/java/beans/menus/EntityDetailResponse.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        this.headers = headers;
+    }
+
+    private void processStyles(Detail detail) {
+        DetailField[] fields = detail.getFields();
+        styles = new Style[fields.length];
+        int i = 0;
+        for (DetailField field : fields) {
+            Style style = new Style(field);
+            styles[i] = style;
```
Comment (benrudolph) [inline src/main/java/beans/menus/EntityDetailResponse.java]: no `append` method for `Style`

File: src/main/java/beans/menus/EntityListResponse.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        processStyles(shortDetail);
+        processActions(shortDetail, nextScreen.getSession());
+    }
+
+    private void processTitle(SessionWrapper session) {
+        setTitle(SessionUtils.getBestTitle(session));
+    }
+
+    private void processEntities(EntityScreen screen, Vector<TreeReference> references, EvaluationContext ec) {
+        entities = new Entity[references.size()];
```
Comment (benrudolph) [inline src/main/java/beans/menus/EntityListResponse.java]: where is this variable declared (might be missing something obvious)? also it doesn't seem to be used

File: src/main/java/beans/menus/EntityListResponse.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        DetailField[] fields = detail.getFields();
+        Object[] data = new Object[fields.length];
+
+        int i = 0;
+        for (DetailField field : fields) {
+            Object o;
+            try {
+                o = field.getTemplate().evaluate(context);
+            } catch(XPathException e) {
+                o = "error (see output)";
```
Comment (benrudolph) [inline src/main/java/beans/menus/EntityListResponse.java]: don't know how this error will manifest for the user, but something more useful might be better

File: src/main/java/beans/menus/EntityListResponse.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        Vector<Action> actions = session.getDetail(((EntityDatum)session.getNeededDatum()).getShortDetail()).getCustomActions();
+        // Assume we only have one TODO WSP: is that correct?
+        if(actions != null && !actions.isEmpty()) {
+            Action action = actions.firstElement();
+            setAction(new DisplayElement(action, session.getEvaluationContext()));
+        }
+    }
+
+    public Entity[] getEntities() {
+        return entities;
```
Comment (benrudolph) [inline src/main/java/beans/menus/EntityListResponse.java]: should this be `this.entities`?

Comment (benrudolph) [inline src/main/java/beans/menus/EntityListResponse.java]: nvm

File: src/main/java/beans/menus/EntityListResponse.java (hunk 8 above)
Comment (benrudolph) [inline src/main/java/beans/menus/EntityListResponse.java]: nvm

File: src/main/java/beans/menus/Style.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    private String widthHint;
+
+    public Style(){}
+
+    public Style(DetailField detail){
+        if(detail.getFontSize() != null) {
+            try {
+                setFontSize(Integer.parseInt(detail.getFontSize()));
+            } catch (NumberFormatException nfe) {
+                nfe.printStackTrace();
```
Comment (benrudolph) [inline src/main/java/beans/menus/Style.java]: no other action is needed if this fails?

File: src/main/java/session/MenuSession.java [hunk 12]
```diff
@@ -60,22 +63,27 @@
...
     private String currentSelection;
 
-    public MenuSession(String username, String password, String domain,
+    Log log = LogFactory.getLog(MenuSession.class);
+
+    public MenuSession(String username, String password, String domain, String appId,
                        String installReference, String serializedCommCareSession,
                        RestoreService restoreService, String sessionId, String currentSelection, InstallService installService) throws Exception {
-        String domainedUsername = StringUtils.getFullUsername(username, domain, host);
+        //TODO WSP: why host isn't host resolving?
```
Comment (benrudolph) [inline src/main/java/session/MenuSession.java]: is this legit?

**Comment by benrudolph** [comment]
looking good. test failure. would you hate me if i asked you to run it through the linter?

File: gradlew (hunk 1 above)
Comment (wpride) [inline gradlew]: This was auto generated from the gradle upgrade I think

File: src/main/java/application/MenuController.java (hunk 2 above)
Comment (wpride) [inline src/main/java/application/MenuController.java]: Curious what you think the distinction should be - in my mind both these things would only be seen by us so didn't bother much to distinguish. Is it "things you'd always want to see vs. things you'd only be interested in when debugging" ?

File: src/main/java/auth/BasicAuth.java (hunk 5 above)
Comment (wpride) [inline src/main/java/auth/BasicAuth.java]: Agreed, will try to stick to it

File: src/main/java/beans/menus/EntityListResponse.java (hunk 9 above)
Comment (wpride) [inline src/main/java/beans/menus/EntityListResponse.java]: Yeah I think we should actually just fail hard here

File: src/main/java/beans/menus/Style.java (hunk 11 above)
Comment (wpride) [inline src/main/java/beans/menus/Style.java]: Yeah I think this is reasonable to ignore

File: src/main/java/session/MenuSession.java (hunk 12 above)
Comment (wpride) [inline src/main/java/session/MenuSession.java]: Yes, this is something I need to fix before merge

**Comment by wpride** [comment]
@benrudolph Not at all, I need someone to make me do it ( @phillipm )

I'll lint the top level PR though so I don't have to deal with retroactive merging

**Comment by benrudolph** [comment]
blocked on java version upgrade on servers

**Pull request merged**